### PR TITLE
portal: Don't encrypt if we deem key to be weak

### PR DESCRIPTION
**Pull request by sophie-h**
Minima are the defaults used by libsecert.

### CODE REVIEW

File: src/portal/error.rs [hunk 1]
```diff
@@ -116,3 +131,25 @@ impl std::fmt::Display for InvalidItemError {
         )
     }
 }
+#[derive(Debug, Clone)]
+pub enum WeakKeyError {
```
Comment (bilelmoussaoui) [inline src/portal/error.rs:135]: Could use a bit of documentation, at least the type itself

File: src/portal/item.rs [hunk 2]
```diff
@@ -93,6 +93,8 @@ impl Item {
     }
 
     pub(crate) fn encrypt(&self, key: &Key) -> Result<EncryptedItem, Error> {
+        key.strength()?;
```
Comment (bilelmoussaoui) [inline src/portal/item.rs:96]:
```suggestion
        key.check_strength()?;
```
or something similar maybe?

File: src/key.rs [hunk 3]
```diff
@@ -29,24 +29,43 @@ static DH_PRIME: Lazy<BigUint> = Lazy::new(|| {
...
 
 impl AsMut<[u8]> for Key {
     fn as_mut(&mut self) -> &mut [u8] {
-        &mut self.0
+        &mut self.key
     }
 }
 
 impl Key {
+    pub fn new(key: Vec<u8>) -> Self {
```
Comment (bilelmoussaoui) [inline src/key.rs:51]: Does it make sense to keep this public?

File: src/key.rs [hunk 4]
```diff
@@ -29,24 +29,43 @@ static DH_PRIME: Lazy<BigUint> = Lazy::new(|| {
...
+        Self::new_with_strength(key, Err(portal::WeakKeyError::StrengthUnknown))
+    }
+
+    pub fn strength(&self) -> Result<(), portal::WeakKeyError> {
```
Comment (bilelmoussaoui) [inline src/key.rs:55]: Should this be public?

**Comment by bilelmoussaoui** [comment]
I took the liberty to rebase this branch btw

File: src/portal/error.rs (hunk 1 above)
Comment (sophie-h) [reply src/portal/error.rs:137]: Added the reasons why we do those checks as well.

File: src/portal/item.rs (hunk 2 above)
Comment (sophie-h) [reply src/portal/item.rs]: Yeah, that explains more about what it's for.

File: src/key.rs (hunk 4 above)
Comment (sophie-h) [reply src/key.rs]: (crate) now

**Review by bilelmoussaoui — APPROVED** [review]

**Pull request merged**